amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+
+
+    Parameters
+    ----------
+    ts_str
+        Any valid string repersenting a time signature. See examples below.
+    enforce_2s_3s
+        Map 4 to 2+2 and 6 to 3+3 etc. following the conventions of those time signatures.
+    minimum_pulse
+        Recursively create further down to the denominator level specified. Defaults to 64 for 64th notes.
+
+    Returns
+    -------
+    list
+        Returns a list of lists with start positions by level.
+
+    Examples
+    --------
+    The following examples index the most interesting level:
+
+    >>> starts_from_time_signature("4/4")[1]  # note the half cycle division
+    [0.0, 2.0, 4.0]
+
+    >>> starts_from_time_signature("4/4", enforce_2s_3s=False)[1]  # note the absence of a half cycle division
+    [0.0, 1.0, 2.0, 3.0, 4.0]
+
+    >>> starts_from_time_signature("2/2")[1]  # note the presence of a half cycle division
+    [0.0, 2.0, 4.0]
+
+    >>> starts_from_time_signature("6/8")[1]  # note the macro-beat division
+    [0.0, 1.5, 3.0]
+
+
+    Numerators like 5 and 7 are supported.
+    Use the total value only to avoid segmentation about the denominator level:
+
+    >>> starts_from_time_signature("5/4")[1]  # note no 2+3 or 3+2 level division
+    [0.0, 1.0, 2.0, 3.0, 4.0, 5.0]
+
+    Or use numerator addition in the form "X+Y/Z" to clarify this level ...
+
+    >>> starts_from_time_signature("2+3/4")[1]  # note the 2+3 division
+    [0.0, 2.0, 5.0]
+
+    >>> starts_from_time_signature("2+2+3/8")[1]  # note the 2+2+3 division
+    [0.0, 1.0, 2.0, 3.5]
+
+    >>> starts_from_time_signature("2+2+2+3/8")[1]  # note the 2+2+2+3 division
+    [0.0, 1.0, 2.0, 3.0, 4.5]
+
+    Only standard denominators are supported:
+    i.e., time signatures in the form X/ one of
+    1, 2, 4, 8, 16, 32, or 64.
+    No so-called "irrational" meters yet (e.g., 2/3), sorry!
+
+    Likewise, the minimum_pulse must be set to one of these values
+    (default = 64 for 64th note) and not be longer than the meter.
+
+    Finally, although we support and provide defaults for time signatures in the form "2+3/8",
+    there is no such support for more than one "/" (i.e., the user must build cases like "4/4 + 3/8" explicitly).
+    """
+
+    # Prep and checks
+    numerator, denominator = ts_str.split("/")
+    numerators = [
+        int(x) for x in numerator.split("+")
+    ]  # Note: support "+" before one "/", e.g., `2+3/4`
+    denominator = int(denominator)
+
+    supported_denominators = [1, 2, 4, 8, 16, 32, 64]
+    if denominator not in supported_denominators:
+        raise ValueError(
+            "Invalid time signature denominator; chose from one of:"
+            f" {supported_denominators}."
+        )
+    if minimum_pulse not in supported_denominators:
+        raise ValueError(
+            "Invalid minimum_pulse: it should be expressed as a note value, from one of"
+            f" {supported_denominators}."
+        )
+
+    # Prep. "hiddenLayer"/s
+    hidden_layer_mappings = (
```

Comment:
Can this be written more clearly? Perhaps it would suffice to comment what this data structure is and what it's being used for.

Reply:
Yup good shout. 

I'll also move it out of the function.

Actually, come to think of it, that's another answer to the debate above. The user could create these mappings. I'll implement that, that's probably enough for a separate commit ...